Exclude user provider configuration for contributors list provider

wsmirnow opened the pull request:
On Opencast systems with enabled LTI reference users, the performance of filters.json and metadata.json admin UI request is drastically decreased. Per default the result contains all known users names as option for contributors and presenters metadata fields. With this patch you have an option to deactivate user listings from some or all user providers.

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

wsmirnow committed:
exclude user provide configuration for contributors list provider

On Opencast systems with enabled LTI reference users, the performance of filters.json and metadata.json admin UI request is drastically decreased. Per default the result contains all known users names as option for contributors and presenters metadata fields. With this patch you have an option to deactivate user listings from some or all user providers.

lkiesow commented on etc/org.opencastproject.index.service.resources.list.provider.ContributorsListProvider.cfg:
```diff
@@ -0,0 +1,14 @@
+# A comma separated list of user provider names to exclude the users from the result set.
+# The available user provider names depends on your system configuration.
+# The may be incomplete list of provider names:
```
```suggestion
# Common user providers are:
```

wsmirnow replied:
suggestion committed

lkiesow commented on etc/org.opencastproject.index.service.resources.list.provider.ContributorsListProvider.cfg:
```diff
@@ -0,0 +1,14 @@
+#   - canvas: Canvas user provider
+#   - moodle: Moodle referenced users
+#   - sakai: Sakai referenced users
+# A special value to match all available providers is: *. Prefere to use this value if you want to exclude all user
+# providers. It will improve performance a lot.
```
```suggestion
# A special value to match all available providers is `*`.
# This can improve the performance quite a bit if you do not need the users.
```

wsmirnow replied:
suggestion committed

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/resources/list/provider/ContributorsListProvider.java:
```diff
@@ -65,13 +70,30 @@
     logger.info("Contributors list provider activated!");
   }
 
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
```
You should [use a declarative OSGi service with a `modified` method instead](https://docs.opencast.org/develop/developer/#proposal-log/#osgi-annotations-and-configuration). For a longer explanation take a look at the [webinar recording](https://recordings.rna1.blindsidenetworks.com/opencast/3cdef0a89e99c065d105b3a90fa2f5575d3cb628-1589986116901/capture/) or the [accompanying slides](https://docs.google.com/presentation/d/10iLzun3JcXfAEihOy5uxDKhyN0xtuYFb1dlBJcOPP4o/edit?usp=sharing) about configuration of OSGi services.

wsmirnow replied:
In general I'm agree with this suggestion. But then it will be the first and only one micro service in the index module with OSGI annotation.

lkiesow replied:
That's not what I meant. I get that introducing annotations here may be a problem (though we should if we can). I'm talking about using `modified`. That's something you can also specify in the XML code. Just set `modified="modified"` just like you would maybe set `activate="activate"`.

The problem with manually implemented managed services is that the configuration is loaded multiple times and in combination with a missing `modified`, the service will be restarted with any configuration change.

wsmirnow replied:
Thanks, suggestion is now implemented.

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/resources/list/provider/ContributorsListProvider.java:
```diff
@@ -65,13 +70,30 @@
 
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
+    if (properties == null) {
+      logger.debug("No configuration available, using defaults.");
```
This will not load the defaults if you delete the configuraion file.

wsmirnow replied:
suggestion committed

lkiesow reviewed (changes requested):
Generally looks good and is definitely an improvement.
Only the configuration loading should be using `modified` instead of manually implementing a managed service.

wsmirnow commented:
I've squashed and rebased the commits.

lkiesow committed:
Ensure configuration in loaded on startup

Starting up a bundle, the activate method is called. Modified is used
only when the configuration is modified during runtime. This ensures the
configuration is applied regardless of which method is executed.

lkiesow reviewed (approved):
The initial reading of the configuration wasn't working properly but fixing that was a one-liner and I just did that. Can you check if that works for you?

The patch looks good and should definitely help. +1 for merging this.

wsmirnow commented:
Yes, thanks.

CGreweling commented:
I have backportet this in our 8.x production system it Works!
A side note: 
Excluding only lti-reference did not make a difference at our system. Using exclude everything ``` * ``` worked very well.
From about 1min load to 1s.

wsmirnow commented:
Sounds good. We have to bring it in before 9.5 release cut imo.

lkiesow commented:
>  Excluding only lti-reference did not make a difference at our system. Using exclude everything * worked very well.

That actually doesn't sound good. It sounds like requesting the user is actually the bottleneck and instead of filtering them later, we should consider not requesting them in the first palce. Oherwise, making users select providers is somewhat confusing and we could also just add a boolean option to turn it on and off.

wsmirnow commented:
> > Excluding only lti-reference did not make a difference at our system. Using exclude everything * worked very well.
> 
> That actually doesn't sound good. It sounds like requesting the user is actually the bottleneck and instead of filtering them later, we should consider not requesting them in the first palce. Oherwise, making users select providers is somewhat confusing and we could also just add a boolean option to turn it on and off.

In my opinion we should exclude the users from contributors list at all. In most cases this feature isn't used as I know. Presenting previously entered values (from ES index) should be enough.

gregorydlogan commented:
@lkiesow Any further objections to this?

gregorydlogan commented:
Per discussion at the technical meeting, merging this.
